### PR TITLE
Add buffer-size cap and Python/Rust boundary property tests (#82, #90)

### LINKED ISSUE #82: Rust: proptest boundary tests for validate_buffer_size (rust/cuprum-rust/src/lib.rs)
## Overview

Add proptest property-based tests for `validate_buffer_size` (`rust/cuprum-rust/src/lib.rs`, line 119).

## Baseline

No Rust-level `#[test]`, proptest, Kani, or Verus coverage currently exists under `rust/cuprum-rust`.

## Justification

Current tests hit `0` and normal values but not platform boundary values such as `i64::MAX` or values that overflow `usize`.

## Benefit

Proves all non-positive values are rejected and accepted values fit `usize` on the target platform.

## Cost Trade-Off

Low.

## Verifiability Improvement

Extract `checked_buffer_size(i64) -> Result<usize, ErrorKind>` and add an explicit upper memory cap.

## Tool Fit

proptest — straightforward numeric boundary coverage.

---
Raised from PR https://github.com/leynos/cuprum/pull/62 by @leynos.

### LINKED ISSUE #90: Rust: proptest at Python/Rust boundary for rust_pump_stream / rust_consume_stream (rust/cuprum-rust/src/lib.rs)
## Overview

Add proptest-driven tests at the Python/Rust boundary for `rust_pump_stream` and `rust_consume_stream` (`rust/cuprum-rust/src/lib.rs`, line 49).

## Baseline

No Rust-level `#[test]`, proptest, Kani, or Verus coverage currently exists under `rust/cuprum-rust`. Current coverage is Python integration/scenario tests only.

## Justification

Error mapping and default argument equivalence are currently scenario-tested only. Invalid conversions and I/O failure paths are not systematically exercised.

## Benefit

- Proves invalid conversion returns `ValueError`
- Proves I/O failures map to `OSError`
- Proves defaults match explicit values

## Cost Trade-Off

Medium. Cross-language tests are slower and require a PyO3 test harness.

## Verifiability Improvement

Add an internal `RustStreamError` enum and convert to PyO3 errors in one place.

## Tool Fit

proptest at the Python/Rust boundary.

## Recommended Order

Phase 2 — after pure Rust unit coverage is established.

---
Raised from PR https://github.com/leynos/cuprum/pull/62 by @leynos.

### PR DESCRIPTION
**Pull request by leynos**
## Summary

Hardens `buffer_size` validation with an explicit upper cap and adds property-based coverage on both sides of the Python/Rust boundary.

### Rust (#82)
- Extract a pure `checked_buffer_size(i64) -> Result<usize, &'static str>` from `validate_buffer_size`, and add an explicit **1 GiB** cap (`MAX_BUFFER_SIZE`). An absurd `buffer_size` is now rejected before allocation instead of attempting a huge `Vec`. The PyO3 wrapper maps the stable message to `PyValueError` as before.
- `buffer_size_tests.rs`: proptest that `checked_buffer_size` accepts exactly the positive, in-range, at-or-below-cap values and rejects non-positive, over-cap, and `usize`-overflowing inputs, with explicit boundary unit tests (1, cap, cap+1, `i64::MAX`, 0, `i64::MIN`) and a drift guard tying the test cap constant to `MAX_BUFFER_SIZE`.

### Python (#90)
- `test_rust_streams_boundary_property.py`: Hypothesis properties at the seam proving non-positive and over-cap `buffer_size` values and negative / out-of-`i32`-range descriptors raise `ValueError`, and that omitting `buffer_size` matches the explicit 65536 default. (Existing example tests already cover `buffer_size=0` and the OSError I/O-failure path.)

Regenerated the maturin wheel-manifest snapshot for the new test module.

## Verification

`make check-fmt lint typecheck markdownlint` green (clippy, whitaker, `cargo doc -D warnings`); `cargo nextest` 37/37. The new Python boundary tests were verified against a locally-built native extension (they, like the existing `rust_streams` suite, skip when the extension is not installed and run in CI). Two unrelated `test_folded_summary.py` Hypothesis-deadline tests flaked under concurrent machine load and pass in isolation.

Closes #82
Closes #90

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: cuprum/unittests/test_rust_streams_boundary_property.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    """``rust_pump_stream`` rejects a non-positive ``buffer_size``."""
+    with pytest.raises(ValueError, match="buffer_size"):
+        rust_streams.rust_pump_stream(_UNUSED_FD, _UNUSED_FD, buffer_size=bad_size)
+
+
+@_SUPPRESS_FIXTURE
+@given(
+    bad_fd=st.one_of(
+        st.integers(min_value=-(1 << 62), max_value=-1),
+        st.integers(min_value=_I32_MAX + 1, max_value=_I64_MAX),
```
[the remark this answers is not in the document]
Comment (pandalump) [reply cuprum/unittests/test_rust_streams_boundary_property.py:107]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please reply confirming this. Otherwise, please provide a follow-up recommendation.

File: rust/cuprum-rust/src/lib.rs [hunk 2]
```diff
@@ -135,19 +137,42 @@ fn convert_platform_fd(value: i64) -> Result<PlatformFd, &'static str> {
...
+fn checked_buffer_size(buffer_size: i64) -> Result<usize, &'static str> {
     if buffer_size <= 0 {
-        return Err(PyValueError::new_err(
-            "buffer_size must be greater than zero",
-        ));
+        return Err("buffer_size must be greater than zero");
+    }
+    let size = usize::try_from(buffer_size).map_err(|_| "buffer_size is too large")?;
+    if size > MAX_BUFFER_SIZE {
+        return Err("buffer_size exceeds the maximum permitted size");
```
[the remark this answers is not in the document]
Comment (pandalump) [reply rust/cuprum-rust/src/lib.rs:162]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please reply confirming this. Otherwise, please provide a follow-up recommendation.

**Comment by buzzybee-df12** [comment]
@coderabbitai review

**Comment by pandalump** [comment]
@coderabbitai On the "Out of Scope Changes" pre-merge check for `.codescene/code-health-rules.json`:

This override is not unrelated to #82 — it is required *because* of the #82 change. Adding `checked_buffer_size(i64)` to `rust/cuprum-rust/src/lib.rs` pushed that module's primitive-argument ratio to 35.5% (> the 30% threshold), tripping CodeScene's Primitive Obsession rule and failing this PR's own "CodeScene Code Health Review" CI gate. The override documents why those raw `i64` FFI inputs are the untyped values the boundary exists to validate into domain newtypes, and is scoped to `lib.rs` only. Without it, this PR cannot go green, so it belongs with the change that caused the finding rather than in a separate PR. The Code Duplication rule_set has been narrowed to the exact file (`io_utils/tests.rs`) per the inline review.

**Comment by pandalump** [comment]
@coderabbitai On the "User-Facing Documentation" and "Developer Documentation" pre-merge checks for the buffer-size cap:

The new 1 GiB `buffer_size` ceiling and its `ValueError` are already documented where this contract lives: the `cuprum/_streams_rs.py` wrapper docstrings (`rust_pump_stream` / `rust_consume_stream`) and `docs/cuprum-design.md` (both the pump and consume signatures), updated in this PR. The `rust_*_stream` helpers are an internal optimization boundary — `rust_consume_stream` is not even wired into production (ADR-002, Phase 2) — so they are deliberately not surfaced in `docs/users-guide.md`, which covers end-user command behaviour. No user-facing behaviour changed. If a dedicated developers-guide section for the Rust stream boundary is wanted as a follow-up, I'm happy to open an issue for it, but it's disproportionate to this test/cap PR.

**Comment by buzzybee-df12** [comment]
@coderabbitai review

**Comment by buzzybee-df12** [comment]
@coderabbitai review

**Comment by pandalump** [comment]
@coderabbitai review

**Comment by pandalump** [comment]
@coderabbitai Both documentation checks are already satisfied on the current branch head (commit `9aac6b9`) — the walkthrough appears to be evaluating a superseded commit:

- **User-Facing Documentation**: `docs/users-guide.md`, under "Rust stream pump (internal)", now states: "Both `rust_pump_stream` and `rust_consume_stream` accept an optional `buffer_size` (bytes, default 64 KiB). It must be a positive integer no larger than 1 GiB (`1 << 30`); a value below 1 or above the cap raises `ValueError`."
- **Developer Documentation**: `docs/developers-guide.md` has a new "Rust stream buffer-size validation" section documenting `checked_buffer_size` / `MAX_BUFFER_SIZE` (1 GiB), the pure-helper seam, and where the contract is property tested.

Could you re-review against head `9aac6b9`? Both docs were added in the "Document the 1 GiB buffer_size cap in the user and dev guides" commit.

**Comment by pandalump** [comment]
@coderabbitai The pre-merge "Developer Documentation" warning recommends marking roadmap items 6.1.1 and 6.1.2 as done. Please reconsider: this PR intentionally implements only a bounded slice of those items, so leaving them unchecked is correct.

Scope of this PR (head `9aac6b9`):
- Adds a **buffer-size cap**: `checked_buffer_size(i64) -> Result<usize, &'static str>` plus `MAX_BUFFER_SIZE = 1 << 30`, wired through `validate_buffer_size` so an invalid `buffer_size` raises `ValueError`.
- Adds boundary property tests for that cap.

Roadmap 6.1.1 requires more than this slice: *"Introduce a `RustStreamError` enum in `rust/cuprum-rust/src/lib.rs` and convert it to PyO3 errors at a single boundary point"*, with the success criterion that **both** invalid `buffer_size` → `ValueError` **and** I/O failures → `OSError` are produced through **one centralized conversion**. This PR does not introduce a `RustStreamError` enum and does not centralize the `OSError` side — you can confirm via `git grep RustStreamError origin/rust-buffer-boundary-tests` returns nothing. 6.1.2 likewise requires the full boundary error-mapping proptest parity, not only the buffer-size cap.

Checking 6.1.1/6.1.2 would therefore overstate completion and make the roadmap inaccurate. The enum-based centralized error mapping is tracked as future work. Could you confirm this reasoning and dismiss the warning?

File: cuprum/unittests/test_rust_streams_boundary_property.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+from __future__ import annotations
+
+import os
+import typing as typ
+
+import pytest
+from hypothesis import HealthCheck, given, settings
+from hypothesis import strategies as st
+
+from tests.helpers.stream_pipes import _safe_close
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/unittests/test_rust_streams_boundary_property.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: rust/cuprum-rust/src/buffer_size_tests.rs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        } else if let Ok(size) = usize::try_from(value) {
+            if size <= MAX_BUFFER_SIZE {
+                prop_assert_eq!(result, Ok(size));
+            } else {
+                prop_assert!(result.is_err(), "{value} above cap must be rejected");
+            }
+        } else {
+            prop_assert!(result.is_err(), "{value} beyond usize must be rejected");
+        }
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply rust/cuprum-rust/src/buffer_size_tests.rs:65]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: docs/users-guide.md [hunk 5]
```diff
@@ -1201,6 +1201,10 @@ The Rust extension now includes an internal pump function exposed as
 Cuprum's internal pipeline dispatcher and may change without notice. Public
 command execution remains unchanged until the dispatcher integration lands.
 
+Both `rust_pump_stream` and `rust_consume_stream` accept an optional
+`buffer_size` (bytes, default 64 KiB). It must be a positive integer no larger
+than 1 GiB (`1 << 30`); a value below 1 or above the cap raises `ValueError`.
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/users-guide.md:1212]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: rust/cuprum-rust/src/buffer_size_tests.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (pandalump) [reply rust/cuprum-rust/src/buffer_size_tests.rs:65]: Addressed in `5e250cb`. Added a reference-oracle helper outside `proptest!`:

```rust
fn expected_buffer_size_result(value: i64) -> Result<usize, &'static str> {
    if value <= 0 {
        return Err("buffer_size must be greater than zero");
    }
    let size = usize::try_from(value).map_err(|_| "buffer_size is too large")?;
    if size > MAX_BUFFER_SIZE {
        return Err("buffer_size exceeds the maximum permitted size");
    }
    Ok(size)
}
```

`matches_bounds` now compares the value and the stable message directly:
`prop_assert_eq!(checked_buffer_size(value), expected_buffer_size_result(value));`.
The existing boundary tests are unchanged. `make check-fmt`, `make lint`, and
`make test` (nextest 38/38, incl. `matches_bounds`) all pass.

@coderabbitai Has this now been resolved in the latest commit? Use codegraph analysis to determine your answer. If this comment is now resolved, please reply confirming this. Otherwise, please provide a follow-up recommendation.

File: docs/users-guide.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (pandalump) [reply docs/users-guide.md:1212]: Addressed in `5e250cb`, limited to `docs/users-guide.md`. The paragraph now preserves the contract (positive integer, max 1 GiB `1 << 30`, 64 KiB default) and qualifies `ValueError` precisely: it is raised after PyO3 converts the argument to a signed 64-bit integer and Rust validation runs, whereas a non-integer or a Python integer outside the i64 range may fail earlier during PyO3 argument conversion with a different exception. No specific conversion exception type is claimed, since that is not verified for the pinned PyO3 version. Prose stays wrapped at 80 columns; `make markdownlint` and `make nixie` pass.

@coderabbitai Has this now been resolved in the latest commit? Use codegraph analysis to determine your answer. If this comment is now resolved, please reply confirming this. Otherwise, please provide a follow-up recommendation.
